src/main/java/edu/kit/compiler/semantic/DetailedNameTypeAstVisitor.java

```diff
@@ -0,0 +1,505 @@
+            Definition definition = symboltable.lookup(localVariableDeclarationStatementNode.getName());
+            if (definition.getKind() != DefinitionKind.Field) {
+                semanticError(localVariableDeclarationStatementNode, "variable is already defined in current scope");
+            }
+        }
+
+        if (localVariableDeclarationStatementNode.getExpression().isPresent()) {
+            DataType rightSideType = localVariableDeclarationStatementNode.getExpression().get().accept(this);
+
+            if (!leftSideType.isCompatibleTo(rightSideType)) {
+                semanticError(localVariableDeclarationStatementNode,
+                    "invalid assigment, variable type is %s while expression type is %s",
+                    leftSideType.getRepresentation(stringTable),
+                    rightSideType.getRepresentation(stringTable)
+                );
+            }
+        }
+
+        symboltable.insert(localVariableDeclarationStatementNode);
+
+        return VoidType;
+    }
+
+    public DataType visit(IfStatementNode ifStatementNode) {
+        DataType conditionType = ifStatementNode.getCondition().accept(this);
+        if (conditionType.getType() != DataTypeClass.Boolean) {
+            semanticError(ifStatementNode, "if statement condition must be boolean");
+        }
+
+        ifStatementNode.getThenStatement().accept(this);
+        if (ifStatementNode.getElseStatement().isPresent()) {
+            ifStatementNode.getElseStatement().get().accept(this);
+        }
+
+        return VoidType;
+    }
+
+    public DataType visit(WhileStatementNode whileStatementNode) {
+        DataType conditionType = whileStatementNode.getCondition().accept(this);
+        if (conditionType.getType() != DataTypeClass.Boolean) {
+            semanticError(whileStatementNode, "while statement condition must be boolean");
+        }
+
+        whileStatementNode.getStatement().accept(this);
+
+        return VoidType;
+    }
+
+    public DataType visit(ReturnStatementNode returnStatementNode) {
+        if (returnStatementNode.getResult().isPresent()) {
+            DataType resultType = returnStatementNode.getResult().get().accept(this);
+
+            if (!resultType.isCompatibleTo(expectedReturnType)) {
+                semanticError(returnStatementNode,
+                    "expression type %s does not match required return type %s",
+                    resultType.getRepresentation(stringTable),
+                    expectedReturnType.getRepresentation(stringTable)
+                );
+            }
+        } else if (expectedReturnType.getType() != DataTypeClass.Void) {
+            semanticError(returnStatementNode, "method requires a return value of type %s", expectedReturnType.getRepresentation(stringTable));
+        }
+
+        return VoidType;
+    }
+
+    public DataType visit(ExpressionStatementNode expressionStatementNode) {
+        expressionStatementNode.getExpression().accept(this);
```

Comment:
Only method invocations and assignments are permitted in _ExpressionStatements_, should this be checked in your phase?

> _ExpressionStatements_, die keine Methodenaufrufe oder Zuweisungen sind, müssen
in der Semantikphase abgewiesen werden. Insbesondere haben _NewObjectExpressions_
in MiniJava keine Seiteneffekte und werden daher ebenfalls abgewiesen.

Reply:
Also my job :D